File: screen_locker/_manual_workout.py
"""Manual (unverified) workout rate-limiting, validation, and entry-building.

Pure logic — no Tk imports. Unlike sick days, manual-workout entries live
directly in ``log.json`` (not a separate history file): a manual
entry is just another ``workout_data`` dict with ``type="manual_workout"``,
so ``LogMixin.save_workout_log`` already HMAC-signs it for free and the
weekly-minimum/bonus logic already counts it once wired into
``COUNTED_WORKOUT_TYPES``. This module only adds the rate-budget check (by
scanning ``log.json`` itself) and the evidence-form validation.

Each sport has its own structured fields (e.g. table tennis tracks
matches/sets won-lost and racket/balls, rather than one free-text blob) —
see :class:`ManualWorkoutDraft` and :data:`SPORT_CHOICES`.
"""

from __future__ import annotations

from datetime import UTC, datetime
import logging

from screen_locker._constants import (
    MANUAL_WORKOUT_DESCRIPTION_MIN_CHARS,
    MANUAL_WORKOUT_REFLECTION_MIN_CHARS,
    MANUAL_WORKOUT_RPE_MAX,
    MANUAL_WORKOUT_RPE_MIN,
    MIN_WORKOUT_DURATION_MINUTES,
    WORKOUT_DURATION_ACCEPT_MINUTES,
)
from screen_locker._manual_workout_types import (
    MANUAL_WORKOUT_TYPE,
    SPORT_CHOICES,
    SPORT_LABELS,
    SPORT_OTHER,
    SPORT_TABLE_TENNIS,
    ManualWorkoutDraft,
    budget_summary,
    count_in_window,
    is_budget_exhausted,
)

__all__ = [
    "MANUAL_WORKOUT_TYPE",
    "SPORT_CHOICES",
    "SPORT_LABELS",
    "SPORT_OTHER",
    "SPORT_TABLE_TENNIS",
    "ManualWorkoutDraft",
    "budget_summary",
    "build_entry",
    "build_sync_payload",
    "count_in_window",
    "is_budget_exhausted",
    "manual_sync_record_id",
    "validate_manual_workout",
]

_logger = logging.getLogger(__name__)


def _parse_hhmm(value: str) -> datetime | None:
    """Parse an ``HH:MM`` string into a datetime on an arbitrary fixed date."""
    try:
        return datetime.strptime(value.strip(), "%H:%M").replace(tzinfo=UTC)
    except ValueError as exc:
        _logger.warning(
            "Manual workout time %r is not HH:MM (%s) — the workout's duration "
            "cannot be computed, so the draft will fail validation",
            value,
            exc,
        )
        return None


def _duration_minutes(draft: ManualWorkoutDraft) -> float | None:
    """Return ``end_time - start_time`` in minutes, or None if unparsable."""
    start = _parse_hhmm(draft.start_time)
    end = _parse_hhmm(draft.end_time)
    if start is None or end is None or end <= start:
        return None
    return (end - start).total_seconds() / 60


def validate_manual_workout(draft: ManualWorkoutDraft) -> str | None:
    """Return an error message if the draft is invalid, else None."""
    if draft.sport not in SPORT_CHOICES:
        return "Please choose a sport"

    required = {
        "Start time": draft.start_time,
        "End time": draft.end_time,
        "Location name": draft.location_name,
        "Transport method": draft.transport_method,
        "Cost": draft.cost,
    }
    for label, value in required.items():
        if not value.strip():
            return f"{label} is required"

    duration = _duration_minutes(draft)
    if duration is None:
        return "Start/end time must be valid HH:MM, with end after start"
    # Accept bar carries the hidden leeway; the message advertises the round
    # number only (see _constants.py).
    if duration < WORKOUT_DURATION_ACCEPT_MINUTES:
        return (
            f"Session must be at least {MIN_WORKOUT_DURATION_MINUTES} "
            f"minutes (currently {duration:.0f})"
        )

    if not MANUAL_WORKOUT_RPE_MIN <= draft.rpe <= MANUAL_WORKOUT_RPE_MAX:
        return (
            f"RPE must be between {MANUAL_WORKOUT_RPE_MIN} and {MANUAL_WORKOUT_RPE_MAX}"
        )

    if draft.sport == SPORT_TABLE_TENNIS:
        error = _validate_table_tennis(draft)
    else:
        error = _validate_other_sport(draft)
    if error is not None:
        return error

    for label, value in (
        ("What went well", draft.went_well),
        ("What to improve", draft.to_improve),
        ("Overall feeling", draft.overall_feeling),
    ):
        if len(value.strip()) < MANUAL_WORKOUT_REFLECTION_MIN_CHARS:
            return (
                f"{label} must be at least {MANUAL_WORKOUT_REFLECTION_MIN_CHARS} "
                f"characters (currently {len(value.strip())})"
            )

    return None


def _validate_table_tennis(draft: ManualWorkoutDraft) -> str | None:
    """Validate the table-tennis-specific fields of a draft."""
    for label, value in (
        ("Matches won", draft.matches_won),
        ("Matches lost", draft.matches_lost),
        ("Sets won", draft.sets_won),
        ("Sets lost", draft.sets_lost),
    ):
        if value < 0:
            return f"{label} cannot be negative"
    if draft.matches_won + draft.matches_lost == 0:
        return "Enter at least one match played (won + lost)"
    if not draft.racket.strip():
        return "Racket is required"
    if not draft.balls.strip():
        return "Balls are required"
    return None


def _validate_other_sport(draft: ManualWorkoutDraft) -> str | None:
    """Validate the generic "other sport" fields of a draft."""
    if not draft.activity_type_other.strip():
        return "Activity type is required"
    if len(draft.activity_details.strip()) < MANUAL_WORKOUT_DESCRIPTION_MIN_CHARS:
        return (
            "What was done must be at least "
            f"{MANUAL_WORKOUT_DESCRIPTION_MIN_CHARS} characters "
            f"(currently {len(draft.activity_details.strip())})"
        )
    return None


def build_entry(draft: ManualWorkoutDraft) -> dict[str, object]:
    """Convert a validated draft into a ``workout_data`` dict for the log."""
    duration = _duration_minutes(draft)
    if draft.sport == SPORT_TABLE_TENNIS:
        activity_label = "table tennis"
    else:
        activity_label = draft.activity_type_other.strip()

    entry: dict[str, object] = {
        "type": MANUAL_WORKOUT_TYPE,
        "source": f"{activity_label} at {draft.location_name.strip()}",
        "sport": draft.sport,
        "activity_type": activity_label,
        "start_time": draft.start_time.strip(),
        "end_time": draft.end_time.strip(),
        "duration_minutes": f"{duration:.1f}" if duration is not None else "",
        "location_name": draft.location_name.strip(),
        "transport_method": draft.transport_method.strip(),
        "cost": draft.cost.strip(),
        "reservation_phone": draft.reservation_phone.strip(),
        "rpe": int(draft.rpe),
        "techniques_practiced": draft.techniques_practiced.strip(),
        "warm_up_minutes": draft.warm_up_minutes.strip(),
        "pain_or_injury": draft.pain_or_injury.strip(),
        "went_well": draft.went_well.strip(),
        "to_improve": draft.to_improve.strip(),
        "overall_feeling": draft.overall_feeling.strip(),
    }
    if draft.sport == SPORT_TABLE_TENNIS:
        entry.update(
            {
                "matches_won": int(draft.matches_won),
                "matches_lost": int(draft.matches_lost),
                "sets_won": int(draft.sets_won),
                "sets_lost": int(draft.sets_lost),
                "racket": draft.racket.strip(),
                "balls": draft.balls.strip(),
            }
        )
    else:
        entry.update(
            {
                "activity_details": draft.activity_details.strip(),
                "equipment": draft.equipment.strip(),
            }
        )
    return entry


# ── Cross-device sync wire format ─────────────────────────────────────────────
# The value of the record ``kind`` discriminator equals ``MANUAL_WORKOUT_TYPE``
# by design: on the wire a manual record is tagged ``kind="manual_workout"`` and
# its payload also carries ``type="manual_workout"``. A StrongLifts session
# payload has no ``kind`` field, so an ingesting PC routes on ``kind`` first and
# a manual record can never be mistaken for (or stamped as) a verified session.
MANUAL_WORKOUT_SYNC_KIND = MANUAL_WORKOUT_TYPE


def manual_sync_record_id(date: str, start_time: str) -> str:
    """Return the stable crdt_sync record id for a manual workout.

    Prefixed ``manual:`` so it never collides with the bare-ISO session ids the
    phone app writes into the same device log. Derived only from
    ``(date, start_time)`` so re-syncing the same workout is idempotent — this
    id must never be re-minted for a given workout.
    """
    return f"manual:{date}T{start_time.strip()}"


def build_sync_payload(draft: ManualWorkoutDraft, date: str) -> dict[str, object]:
    """Build the cross-device sync payload for a manual workout.

    This is the wire contract shared with the phone app: the same fields as
    :func:`build_entry` (what lands in ``log.json``), plus a ``kind``
    discriminator and the workout's own ``date``. ``log.json`` is
    day-keyed on the PC, but the synced payload must carry its date so the
    shared budget can be windowed and the entry filed in the right ISO week on
    ingest.
    """
    payload = build_entry(draft)
    payload["kind"] = MANUAL_WORKOUT_SYNC_KIND
    payload["date"] = date
    return payload

File: stronglift_replacement/workout_app/lib/models/manual_workout.dart
/// Manual (off-app) workout model, validation, and sync-record building.
///
/// Mirrors the Python `screen_locker._manual_workout` module
/// (`ManualWorkoutDraft`, `validate_manual_workout`, `build_entry` /
/// `build_sync_payload`, `manual_sync_record_id`). The payload wire format is
/// pinned by a cross-language fixture shared with the Python side
/// (`test/models/manual_workout_test.dart` and its twin
/// `test_manual_workout.py::TestSyncWireFormat`) — keep the two in lockstep.
library;

import 'package:crdt_sync/crdt_sync.dart';

part 'manual_workout_budget.dart';
part 'manual_workout_sync.dart';
part 'manual_workout_validation.dart';

// ── Shared constants ─────────────────────────────────────────────────────────
// Hand-maintained mirror of the Python source of truth in
// `screen_locker/_constants.py` (the `MANUAL_WORKOUT_*` values).
//
// The payload fixture does NOT pin these: the caps never travel on the wire,
// so it cannot see them drift — and they did (298daf0 raised Python 5 -> 10
// touching no Dart, leaving the phone at 5). The actual tripwire is
// `TestCrossLanguageBudgetConstants` in
// `screen_locker/tests/test_manual_workout_features.py`, which parses THIS
// file and compares it to the Python constants, so either side drifting fails.

/// Max manual workouts allowed in any rolling 7-day window.
const int kManualWorkoutBudgetPer7Days = 2;

/// Max manual workouts allowed in any rolling 30-day window.
const int kManualWorkoutBudgetPer30Days = 10;

/// The ONE duration bar every workout type answers to, in minutes. This is
/// the ADVERTISED number — the only one that may appear in a message the user
/// reads. Mirrors `MIN_WORKOUT_DURATION_MINUTES` in the locker's
/// `_constants.py`; the two must move together.
const int kMinWorkoutDurationMinutes = 40;

/// The leeway that is deliberately INVISIBLE. A session is accepted at
/// [kMinWorkoutDurationMinutes] minus this, so 35 minutes passes while every
/// string still says 40. That gap is intentional: do NOT "fix" it by
/// advertising the real cutoff, and never interpolate
/// [kWorkoutDurationAcceptMinutes] into a user-facing string.
const int kWorkoutDurationLeewayMinutes = 5;

/// Derived so the advertised and accept bars cannot drift apart. Comparisons
/// use this; messages use [kMinWorkoutDurationMinutes]. Compare with `>=`,
/// matching the locker, so 35.0 cannot pass one side and fail the other.
const int kWorkoutDurationAcceptMinutes =
    kMinWorkoutDurationMinutes - kWorkoutDurationLeewayMinutes;

/// Minimum characters for the "other sport" activity description.
const int kManualWorkoutDescriptionMinChars = 40;

/// Minimum characters required in each reflection field.
const int kManualWorkoutReflectionMinChars = 20;

/// Lowest valid RPE (rate of perceived exertion).
const int kManualWorkoutRpeMin = 1;

/// Highest valid RPE (rate of perceived exertion).
const int kManualWorkoutRpeMax = 10;

/// Record `kind` discriminator marking a manual-workout sync record.
///
/// A StrongLifts session payload has no `kind`; the ingesting PC routes on this
/// value before any session logic, so a manual can never be stamped verified.
const String kManualWorkoutSyncKind = 'manual_workout';

/// Internal sport code for table tennis (structured score fields).
const String kSportTableTennis = 'table_tennis';

/// Internal sport code for any other sport (free-text activity fields).
const String kSportOther = 'other';

/// The sport codes a manual workout may use.
const List<String> kSportChoices = [kSportTableTennis, kSportOther];

/// Human labels for the sport codes, in menu order.
const Map<String, String> kSportLabels = {
  kSportTableTennis: 'Table tennis',
  kSportOther: 'Other',
};

/// User-supplied evidence fields for a manual (unverified) workout.
///
/// Fields shared by every sport come first; the sport-specific groups follow —
/// only the group matching [sport] is validated/persisted.
class ManualWorkoutDraft {
  /// Creates a manual-workout draft.
  const ManualWorkoutDraft({
    required this.sport,
    required this.startTime,
    required this.endTime,
    required this.locationName,
    required this.transportMethod,
    required this.cost,
    required this.rpe,
    required this.wentWell,
    required this.toImprove,
    required this.overallFeeling,
    this.reservationPhone = '',
    this.techniquesPracticed = '',
    this.warmUpMinutes = '',
    this.painOrInjury = 'none',
    this.matchesWon = 0,
    this.matchesLost = 0,
    this.setsWon = 0,
    this.setsLost = 0,
    this.racket = '',
    this.balls = '',
    this.activityTypeOther = '',
    this.activityDetails = '',
    this.equipment = '',
  });

  /// Sport code (see [kSportChoices]).
  final String sport;

  /// Start time as `HH:MM`.
  final String startTime;

  /// End time as `HH:MM`.
  final String endTime;

  /// Where the session happened.
  final String locationName;

  /// How the user travelled there.
  final String transportMethod;

  /// Session cost (free text, e.g. `40 PLN`).
  final String cost;

  /// Rate of perceived exertion (1-10).
  final int rpe;

  /// Reflection: what went well.
  final String wentWell;

  /// Reflection: what to improve.
  final String toImprove;

  /// Reflection: overall feeling.
  final String overallFeeling;

  /// Optional reservation phone number.
  final String reservationPhone;

  /// Optional techniques/focus areas practised.
  final String techniquesPracticed;

  /// Optional warm-up duration (free text).
  final String warmUpMinutes;

  /// Pain or injury notes (defaults to `none`).
  final String painOrInjury;

  /// Table tennis: matches won.
  final int matchesWon;

  /// Table tennis: matches lost.
  final int matchesLost;

  /// Table tennis: sets won.
  final int setsWon;

  /// Table tennis: sets lost.
  final int setsLost;

  /// Table tennis: racket used.
  final String racket;

  /// Table tennis: balls used.
  final String balls;

  /// Other sport: what activity it was.
  final String activityTypeOther;

  /// Other sport: description of what was done.
  final String activityDetails;

  /// Other sport: equipment used.
  final String equipment;
}


/// Builds the persisted `workout_data` dict for a validated draft.
///
/// Byte-for-byte equivalent of the Python `build_entry`.
Map<String, Object?> buildEntry(ManualWorkoutDraft draft) {
  final duration = durationMinutes(draft);
  final activityLabel = draft.sport == kSportTableTennis
      ? 'table tennis'
      : draft.activityTypeOther.trim();

  final entry = <String, Object?>{
    'type': kManualWorkoutSyncKind,
    'source': '$activityLabel at ${draft.locationName.trim()}',
    'sport': draft.sport,
    'activity_type': activityLabel,
    'start_time': draft.startTime.trim(),
    'end_time': draft.endTime.trim(),
    'duration_minutes': duration != null ? duration.toStringAsFixed(1) : '',
    'location_name': draft.locationName.trim(),
    'transport_method': draft.transportMethod.trim(),
    'cost': draft.cost.trim(),
    'reservation_phone': draft.reservationPhone.trim(),
    'rpe': draft.rpe,
    'techniques_practiced': draft.techniquesPracticed.trim(),
    'warm_up_minutes': draft.warmUpMinutes.trim(),
    'pain_or_injury': draft.painOrInjury.trim(),
    'went_well': draft.wentWell.trim(),
    'to_improve': draft.toImprove.trim(),
    'overall_feeling': draft.overallFeeling.trim(),
  };
  if (draft.sport == kSportTableTennis) {
    entry.addAll(<String, Object?>{
      'matches_won': draft.matchesWon,
      'matches_lost': draft.matchesLost,
      'sets_won': draft.setsWon,
      'sets_lost': draft.setsLost,
      'racket': draft.racket.trim(),
      'balls': draft.balls.trim(),
    });
  } else {
    entry.addAll(<String, Object?>{
      'activity_details': draft.activityDetails.trim(),
      'equipment': draft.equipment.trim(),
    });
  }
  return entry;
}
